datalad/core/distributed/push.py

```diff
@@ -796,14 +796,20 @@ def _push_data(ds, target, content, data, force, jobs, res_kwargs,
     # to this location
     # XXX must not be a SpooledTemporaryFile -- dunno why, but doesn't work
     # otherwise
+    seen_keys = set()
     with TemporaryFile() as file_list:
         nbytes = 0
         for c in to_transfer:
             file_list.write(
                 bytes(Path(c['path']).relative_to(ds.pathobj)))
             file_list.write(b'\0')
-            nbytes += c['bytesize']
-
+            key = c['key']
+            # Avoid inflated estimate due to files pointing to the same key.
+            if key not in seen_keys:
+                seen_keys.add(key)
+                nbytes += c['bytesize']
+        lgr.debug("Counted %d bytes of annex data to transfer",
+                  nbytes)
```

Comment:
Should may be file_list also not gather paths corresponding to duplicate keys? If anything, it might allow annex to avoid multiple parallel workers to complete to transfer the same key?

Reply:
> Should may be file_list also not gather paths corresponding to duplicate keys? If anything, it might allow annex to avoid multiple parallel workers to complete to transfer the same key?

Good point.

My first thought was that doing so would change the result records that are returned.  Perhaps we'd want to fill in results for those files because it is a bit confusing to silently drop them (especially for paths that are explicitly given) and could be considered a breaking change in terms of result hooks.

To add to the confusion, it looks like by default we already drop the result records for extra files pointing to the same key (at least in some cases).  That's because, with `--fast`, `git annex copy` doesn't emit a record for duplicates (at least in some cases).  To me, that's an additional reason to filter out the seen-key files, and then tack on notneeded results for them at the end.

<details>
<summary>demo</summary>

```sh
set -ue

cd "$(mktemp -d "${TMPDIR:-/tmp}"/dl-push-XXXXXXX)"

datalad create a >/dev/null 2>&1
git -C a config receive.denyCurrentBranch updateInstead

datalad clone a b >/dev/null 2>&1
(
    cd b

    printf 0 >f1
    printf 0 >f2
    printf 0 >f3
    printf 1 >f4
    printf 1 >f5
    printf 1 >f6
    datalad save >/dev/null 2>&1

    echo "=== --fast ========================================="
    printf "f1\0f2\0f3\0" |
        git annex copy --batch --to origin \
            --json --json-error-messages --fast -z \
            | jq -r .file
    echo "=== no --fast ======================================"
    printf "f4\0f5\0f6\0" |
        git annex copy --batch --to origin \
            --json --json-error-messages -z \
            | jq -r .file
)

```

```
=== --fast =========================================
f1
=== no --fast ======================================
f4
f5
f6
```

</details>